Fix #2925: Add ExplorationBackendApiService

joshuacano opened the pull request:
This is the first PR for building an two `ExplorationBackendApiService` services, an editable version and a read_only version.

I decided to open this PR so that I can ensure which E2E tests are failing, (I can't get any more insight on my box as they seem to fail semi-randomly) I have a few outstanding questions to have answered before I ask for a review. 

1. the `deleteExploration` functionality of the `EditableExplorationBackendApiService` what should it return on success and failure?
2. Not sure I have a total handle on which files need the ReadOnly version of the service. Currently this is only in `StatisticsTab.js` and `EditorServices.js`, should there be another setting added to the data returned? I'm assuming this will be handled by the upcoming Objectfactory addressed in another ticket.
3. I used String(explorationId) in 
`
var explorationDataUrl = UrlInterpolationService.interpolateUrl(
          EDITABLE_EXPLORATION_DATA_URL_TEMPLATE, {
            exploration_id: String(explorationId)
          });
` 
To address times when integers might be passed in as an explorationId, I imagine this is not the best solution, and ask if there are any better ideas to be used here.

seanlip commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,162 @@
+ */
+
+oppia.factory('EditableExplorationBackendApiService', [
+    '$http', '$q', 'EXPLORATION_DATA_URL_TEMPLATE',
```
should be 2-space indentation

joshuacano replied:
Gotcha, Have corrected here as well as ReadOnlyExplorationBackendApiService.js

seanlip commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,162 @@
+          });
+        }
+      };
+    }]);
```
NB: you'll know something is wrong when this indentation level doesn't match that of the opening braces.

joshuacano replied:
Understood. Fixed.

seanlip commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,162 @@
+          params.role = JSON.stringify(role);
+        }
+
+        $http.delete(editableExplorationDataUrl, {
```
ah -- this is why the travis production tests are failing. 'delete' is a reserved word in JS. Do something like `$http['delete']` instead.

joshuacano replied:
I see this has been changed.

joshuacano commented:
I was bashing my head against the e2e_tests for a solid day to figure out why it was failing. But https://github.com/oppia/oppia/pull/2958 fixed my issues thankfully! I added a delete test and have it ready to review for @BenHenning, but still have those above 3 questions for Ben when you get a chance.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,163 @@
+    EDITABLE_EXPLORATION_DATA_URL_TEMPLATE, UrlInterpolationService,
+    ReadOnlyExplorationBackendApiService) {
+    var _fetchExploration = function(
+        explorationId, applyDraft, v, successCallback, errorCallback) {
```
What is the ``v`` parameter?

joshuacano replied:
v is version in this case. In a few places its referred to as v and in some version. I'm thinking I will just change this to version.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,163 @@
+        explorationId, applyDraft, v, successCallback, errorCallback) {
+      var explorationDataUrl = UrlInterpolationService.interpolateUrl(
+        EDITABLE_EXPLORATION_DATA_URL_TEMPLATE, {
+          exploration_id: String(explorationId)
```
Per your question, do we need to convert this to a string? Exploration IDs should always be strings.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,163 @@
+      },
+
+      /**
+       * Updates a exploration in the backend with the provided exploration
```
Nit: ``an exploration``

joshuacano replied:
Done.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,163 @@
+       * if one is provided to the returned promise object. Errors are passed
+       * to the error callback, if one is provided. Finally, if the update is
+       * successful, the returned exploration will be cached within the
+       * ExplorationBackendApiService to ensure the cache is
```
Should this be ``ReadOnlyExplorationBackendApiService``?

joshuacano replied:
Correct, done.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,163 @@
+        });
+      },
+
+      deleteExploration: function(
```
Please add documentation comment for this method, too.

joshuacano replied:
Done.

BenHenning replied:
Did you?

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,163 @@
+    };
+
+    return {
+      fetchExploration: function(explorationId, applyDraft, v) {
```
Do we need a fetch, or can we just use ``ReadOnlyExplorationBackendApiService`` in those cases?

joshuacano replied:
Let me delete this and try to use ReadOnlyExplorationBackendApiService as the sole place to load data to start.

joshuacano replied:
For this pass, I kept fetchExploration as a means of loading the applyDraft to the exploration. (It seems this is only for editor Preview mode) Should I remove this and add a loadExplorationWithDraftApplied method to the ReadOnlyHandler?

BenHenning replied:
Let's keep this fetch for now, but make it explicit via documentation comments when to use one or the other.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,163 @@
+      $http['delete'](editableExplorationDataUrl, {
+        params: params
+      }).then(function(response) {
+        // What does this return? What happens on Fail?
```
It doesn't look like this returns anything. You can just have an empty function here, I think, adn not pass anything to the success callback.

joshuacano replied:
Gotcha this has been changed to delete from cache on success and pass nothing to callback. Thanks!

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,249 @@
+      'EditableExplorationBackendApiService');
+    ReadOnlyExplorationBackendApiService = $injector.get(
+      'ReadOnlyExplorationBackendApiService');
+    UndoRedoService = $injector.get('UndoRedoService');
```
Do we actually need this for the test?

joshuacano replied:
No we don't, I believe I was using them in an earlier iteration of a test and never took it out. Removed.

BenHenning replied:
Did you remove it? It still seems to be here.

joshuacano replied:
Not removed. Way to go me. Removed now. Thanks for catching.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,249 @@
+      exploration_id: '0',
+      draft_changes: null,
+      profile_picture_data_url: '',
+      email_preferences: {
```
Is this information actually returnable from the backend? It seems like there's a lot of excess information here that isn't actually returned from the ``ExplorationHandler``.

joshuacano replied:
There is a lot of elements returned from the ExplorationHandler but the real pertinent data is a small subset of this. Also I may have been getting this original info from the ```reader``` handler and not the ```editor``` handler. Will double check and correct.

joshuacano replied:
I've removed a lot of this extra data, let me know if you think I have still included too much information.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,249 @@
+    expect(successHandler).toHaveBeenCalledWith(exploration);
+    expect(failHandler).not.toHaveBeenCalled();
+
+    expect(ReadOnlyExplorationBackendApiService.isCached('0')).toBeTruthy();
```
Prefer: ``toBe(true)`` over ``toBeTruthy()``

joshuacano replied:
Done,good to know.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,249 @@
+    expect(successHandler).toHaveBeenCalledWith({});
+    expect(failHandler).not.toHaveBeenCalled();
+
+    expect(ReadOnlyExplorationBackendApiService.isCached('0')).toBeFalsy();
```
Similarly, prefer ``toBe(false)`` over ``toBeFalsy()``

joshuacano replied:
Done.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiService.js:
```diff
@@ -0,0 +1,131 @@
+
+/**
+ * @fileoverview Service to retrieve read only information
+ * about exploraions from the backend.
```
``explorations``

joshuacano replied:
Done.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiService.js:
```diff
@@ -0,0 +1,131 @@
+
+    return {
+      /**
+       * Retrieves a exploration from the backend given a exploration ID.
```
``an exploration from`` and ``an exploration ID``

joshuacano replied:
Done.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiService.js:
```diff
@@ -0,0 +1,131 @@
+      /**
+       * Retrieves a exploration from the backend given a exploration ID.
+       * This returns a promise object that allows a success and rejection
+       *  callbacks to be registered. If the exploration is successfully
```
``* callbacks to be``

joshuacano replied:
Done.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiService.js:
```diff
@@ -0,0 +1,131 @@
+      },
+
+      /**
+       * Delete a specific exploration from the local cache
```
``Deletes``

joshuacano replied:
Done.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiService.js:
```diff
@@ -0,0 +1,131 @@
+       * passed in as a parameter. If something goes wrong while trying to
+       * fetch the exploration, the rejection callback is called instead,
+       * if present. The rejection callback function is passed the error that
+       * occurred and the exploration ID.
```
Is this last part true? It doesn't seem to pass the exploration ID in the event of an error.

joshuacano replied:
Correct, I have added the exploration ID to the errorCallback, will modify the test to check for this.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,230 @@
+      exploration_id: '0',
+      draft_changes: null,
+      profile_picture_data_url: '',
+      email_preferences: {
```
Ditto here. There seems to be a lot of extra data that's not part of an exploration.

joshuacano replied:
Will take same approach as above.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,230 @@
+    var failHandler = jasmine.createSpy('fail');
+
+    // The exploration should not currently be cached.
+    expect(ReadOnlyExplorationBackendApiService.isCached('0')).toBeFalsy();
```
``toBe(false)``, ditto elsewhere

joshuacano replied:
Done.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiServiceSpec.js:
```diff
@@ -0,0 +1,230 @@
+    expect(failHandler).not.toHaveBeenCalled();
+
+    // The exploration should now be cached.
+    expect(ReadOnlyExplorationBackendApiService.isCached('0')).toBeTruthy();
```
``toBe(true)``, ditto elsewhere

joshuacano replied:
Done.

BenHenning commented on core/templates/dev/head/pages/exploration_editor/history_tab/HistoryServices.js:
```diff
@@ -192,14 +192,14 @@ oppia.factory('compareVersionsService', [
         return $q.all({
-          v1Data: $http.get(explorationDataUrl + v1),
-          v2Data: $http.get(explorationDataUrl + v2)
+          v1Data: EditableExplorationBackendApiService.fetchExploration(
```
Why fetch? Because of the version?

joshuacano replied:
In the current iteration, EditableExplorationBackendApiService only has fetch. I'll be changing this to pull from the ReadOnlyBackendApiService to use fetch for version types.

joshuacano replied:
This has been changed to use ReadOnlyBackendApiService, I have only used EditableExplorationBackendApiService for any instances when we want to use apply_draft = true.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,163 @@
+
+      params = {};
+      if (role) {
+        params.role = JSON.stringify(role);
```
Is this supposed to be a param? Before it was appended to the URL itself.

joshuacano replied:
I was under the impression that params are appended to the url in an angular $.get call, is my impression incorrect? If so I'll change this to just be appended to the url.  i.e.
```if (role) { url += '&role=' + role}```

BenHenning replied:
You can use the URLInterpolationService for this, instead. We should never directly modify URLs except through that service (happy to discuss offline if you have any questions).

BenHenning replied:
Ping. :) We did this with version, so we should also be able to do it with ``role``, right?

joshuacano replied:
COMPLETELY. Doing now.

joshuacano replied:
I have added a few methods to the editableExplorationService, one for fetchApplyDraftExploration as well as adding role as an argument for deleteExploration

joshuacano commented:
Thanks for the comments Ben! I'll get a chance to look and address them this weekend.

joshuacano commented:
One quick question before I address the review comments (Which are very helpful by the way) I don't see anything about the learner view in issue #2925, was there another issue you were referring to?

BenHenning commented:
Yes @joshuacano you have the right idea. We should consolidate all fetching of exploration data to a read-only handler for both the exploration player and editor, and all updating of exploration data to a read-write handler for the exploration editor (I say read because the data needed by the exploration editor may be different than the learner and we may not be able to have just 1 controller for reading both). Does that answer your question?

joshuacano commented:
@BenHenning  It does answer my question. Sorry I didnt respond to you sooner! I'm making good progress on this, but had just one more set of questions:

in editor.py:
We are trying to basically reflect the ExplorationHandler in the EditableExplorationBackendApiService.

Will we be eventually moving the autosave functionality (EditorAutoSaveHandler) and the Revert functionality (ExplorationRevertHandler) into here as well? Is this something I should be doing in here as well? Or is that a future iteration of this issue (along with the domain object work)

@seanlip Also consulting you here in case you might know as well.

seanlip commented:
In general I would recommend splitting functionality into separate PRs where possible, as long as the intermediate result is self-contained and doesn't break anything.

joshuacano commented:
@BenHenning , I have committed my latest changes and am requesting a new review. I think I addressed most of your concerns just a couple of notes.

1. I added a template for ReadOnlyExplorationBackendApiService for loading previous versions. Let me know if thats the best approach. And perhaps I should take that same approach for EditableExplorationBackendApiService. (with apply_draft=true)
2. For now, I am still using the fetchExploration method for EditableExplorationBackendApiService. and checking if apply_draft = true. Another way to do this is to load from ReadOnly anytime apply_draft is null or false. Let me know if the way I have it coded now is sufficient or you think it might be better for me to change the code to use ReadOnly.
3. I added a few items to reader.py, init_state_name, states and exploration_id. From what I can tell these are used in editor.py and using those in reader.py. HOWEVER, the states object is inside of the exploration returned in reader.py. so, my proposal is to either address resolving this in a future PR. Or change the callers for this to use exploration.states returned from the ReadOnlyExplorationHandler instead of just states.

@BenHenning Thanks so much for putting up with me through all this. I feel like we are getting closer!

seanlip commented:
@joshuacano -- there are merge conflicts, can you please fix them? Thanks!

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -58,7 +55,7 @@ oppia.factory('EditableExplorationBackendApiService', [
         UrlInterpolationService.interpolateUrl(
         EDITABLE_EXPLORATION_DATA_URL_TEMPLATE, {
-          exploration_id: explorationId
+          exploration_id: String(explorationId)
```
Why is this necessary?

joshuacano replied:
Currently it's only necessary due to tests in HistoryServiceSpec.js. In line 177, there is the section 
``` 
beforeEach(function() {
      mockExplorationData = {
        explorationId: 0
      };
```
I can change this to a quoted zero and all tests pass. For the version however I must use String to convert from version Number(an int) to a string. I'll go ahead and quote this zero and remove the String conversion for explorationId.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiService.js:
```diff
@@ -14,25 +14,35 @@
+      if (version) {
+        explorationDataUrl = UrlInterpolationService.interpolateUrl(
+        EXPLORATION_VERSION_DATA_URL_TEMPLATE, {
+          exploration_id: String(explorationId),
```
Why is the String conversion necessary for exploration ID?

joshuacano replied:
Currently it's only necessary due to tests in HistoryServiceSpec.js, as mentioned above.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiService.js:
```diff
@@ -14,25 +14,35 @@
-        explorationId, successCallback, errorCallback) {
-      var explorationDataUrl = UrlInterpolationService.interpolateUrl(
+        explorationId, version, successCallback, errorCallback) {
+      var explorationDataUrl = '';
```
Consider moving the if (version) logic to a helper method that returns one option or the other, that way we can directly initialize explorationDataUrl and are never in a state where it's an invalid value.

joshuacano replied:
Yes. This is extremely obvious now, thanks.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiService.js:
```diff
@@ -14,25 +14,35 @@
 
-      $http.get(explorationDataUrl).then(function(response) {
+      $http.get(explorationDataUrl).then(
+        function(response) {
```
Why move this to its own line?

joshuacano replied:
accident, have corrected.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiService.js:
```diff
@@ -75,16 +85,19 @@ oppia.factory('ReadOnlyExplorationBackendApiService', [
+            _fetchExploration(explorationId, version, function(exploration) {
               // Save the fetched exploration to avoid future fetches.
-              _explorationCache[explorationId] = exploration;
+              // Excludes version for now.
```
Please mention this in the documentation for ``loadExploration`` so other developers are aware of this constraint. Should we instead have a ``loadExploration`` and ``loadLatestExploration``? It might be clearer than having an optional version parameter.

joshuacano replied:
I love this idea. Also, I was thinking loadVersionedExploration might be a good name for the versioned load method? But on second thought, the name loadExploration that requires an extra parameter seems logical enough. Done. 

Do we want an additional cache for Versioned Explorations? Or is that overkill for now.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiServiceSpec.js:
```diff
@@ -107,9 +75,9 @@ describe('Read only exploration backend API service', function() {
+    $httpBackend.expect('GET', '/explorehandler/init/0').respond(
       sampleDataResults);
-    ReadOnlyExplorationBackendApiService.fetchExploration('0').then(
+    ReadOnlyExplorationBackendApiService.fetchExploration('0', null).then(
```
Do we have any tests verifying the different behavior when a version is provided? Caching doesn't happen, etc.

joshuacano replied:
Writing up tests for the new method.

BenHenning commented on core/templates/dev/head/pages/exploration_editor/settings_tab/SettingsTab.js:
```diff
@@ -35,9 +35,9 @@ oppia.controller('SettingsTab', [
-      CATEGORY_LIST, explorationAdvancedFeaturesService,
-      EXPLORATION_TITLE_INPUT_FOCUS_LABEL, userExplorationEmailsHandler,
-      EditableExplorationBackendApiService) {
+      explorationAdvancedFeaturesService, ALL_CATEGORIES,
```
Why change these dependencies?

joshuacano replied:
This was just a merge with develop which had changed dependencies. I merely added EditableExplorationBackendApiService

BenHenning commented on core/templates/dev/head/pages/exploration_player/PlayerServices.js:
```diff
@@ -188,39 +190,30 @@ oppia.factory('oppiaPlayerService', [
-            }
-          }).then(function(response) {
-            exploration = ExplorationObjectFactory.create(response.data);
+          EditableExplorationBackendApiService.fetchExploration(
```
This is quite nice. The new services make this if check MUCH more readable!

joshuacano replied:
Thanks! So great to hear that and I agree.

BenHenning commented on feconf.py:
```diff
@@ -487,7 +487,6 @@ def get_empty_ratings():
 DASHBOARD_EXPLORATION_STATS_PREFIX = '/dashboardhandler/explorationstats'
 EDITOR_URL_PREFIX = '/create'
 EXPLORATION_DATA_PREFIX = '/createhandler/data'
-EDITABLE_EXPLORATION_DATA_PREFIX = '/editable_createhandler/data'
```
Was any new code introduced in editor.py also removed?

joshuacano replied:
I believe so, I will double check.

BenHenning commented on core/controllers/reader.py:
```diff
@@ -227,12 +228,22 @@ def get(self, exploration_id):
             'exploration': exploration.to_player_dict(),
+            'exploration_id': exploration_id,
+            'init_state_name': exploration.init_state_name,
+            'states': states,
```
Yeah, I agree with you this really shouldn't be done (even temporarily to be followed up in an upcoming PR). It's adding extra bloat to the payload for every load of the learner view. We probably don't need any of these three if we update endpoints to access the attached exploration, instead.

joshuacano replied:
I agree with you. I'll update the endpoints that I can find which use the reader data and update there. (There are not too many.)

BenHenning commented:
Thanks @joshuacano, this is looking great! I agree that the apply_draft separation makes sense, but we should make it abundantly clear via documentation when to use one service for fetching versus another.

I offered some suggestions for a possible cleaner way to address the versioning dilemma in the read only service. Please add tests for any new functions you introduce.

I made a reply to an earlier comment about the URL params, please take a look.

Finally, I think it's worth addressing the states/exploration_id/init_state_name top-level values in this PR, as having those duplicated is going to drastically increase the size of message data being sent to the client unnecessarily. Please let me know if this ends up being a lot more work than is worth it.

joshuacano commented:
Fantastic. I love the version idea, I'll try and implement these top level ideas and then address the rest of the review. Thanks!

Joshua Cano

> On Mar 7, 2017, at 20:22, Ben Henning <notifications@github.com> wrote:
> 
> Thanks @joshuacano, this is looking great! I agree that the apply_draft separation makes sense, but we should make it abundantly clear via documentation when to use one service for fetching versus another.
> 
> I offered some suggestions for a possible cleaner way to address the versioning dilemma in the read only service. Please add tests for any new functions you introduce.
> 
> I made a reply to an earlier comment about the URL params, please take a look.
> 
> Finally, I think it's worth addressing the states/exploration_id/init_state_name top-level values in this PR, as having those duplicated is going to drastically increase the size of message data being sent to the client unnecessarily. Please let me know if this ends up being a lot more work than is worth it.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

joshuacano commented on core/templates/dev/head/pages/exploration_player/PlayerServices.js:
```diff
@@ -188,39 +190,30 @@ oppia.factory('oppiaPlayerService', [
-            exploration = ExplorationObjectFactory.create(data.exploration);
-            version = data.version;
+          ReadOnlyExplorationBackendApiService.loadExploration(
+            _explorationId, version ? version : null).then(
```
I am wrestling with this. Do I leave this as loadExploration? Or do I add a loadLatestExploration check. I'll go with the latter and see if you (@BenHenning) like the way it looks.

joshuacano commented:
@BenHenning Another question that has come up.

I'm running into a small issue and want to run my solution by you. SO the reader.py and editor.py both return slightly different data, so caching the data between the two proves to be a little troublesome. I was writing out a ConvertEditableToReadOnlyExploration method in the ReadOnlyBackendApiService but quickly realized this: A) was tedious and B) would create unwanted dependencies between the object model and the front end service.
SO my solution is to wipe the cache for this exploration_id anytime an exploration is written by the editableExplorationBackendApiService. At some point it may be nice to convert between these two objects, but I figure/guess that we will create a domain object for a readonly and an editable in the future.
Does this solution sound viable to you?
for Reference here is the response from reader.py:
```
can_edit:true
exploration:Object
exploration_id:"JnWc9t76gkZx"
is_admin:false
is_logged_in:true
is_moderator:false
is_super_admin:false
profile_picture_data_url:"base64:xxxxx"
```

joshuacano commented:
Alright I have pushed my Commits and am asking for another review. @BenHenning I don't think its 100% perfect but its as close as it has been. One major difference is the EditableExplorationHandler returns the exploration with various other attributes directly to the caller (I.e. states is a first level attribute). However, ReadOnly returns 4 additional attributes and then wraps the exploration inside of the attribute exploration (I.e. states is a second level attribute wrapped inside exploration) Let me know if you have any questions at all. Thanks!

BenHenning commented on core/controllers/editor_test.py:
```diff
@@ -342,6 +342,7 @@ def _create_training_data(*arg):
             self.assertEqual(
                 response_dict['unhandled_answers'],
                 _create_training_data('joyful', 'sad'))
+            self.assertTrue(exploration_dict['version'])
```
Do you mean to do: ``self.assertIn('version', exploration_dict)``?

joshuacano replied:
oh I didn't change this. Thats strange, must have been a result of a merge.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,161 @@
+        ReadOnlyExplorationBackendApiService.deleteExplorationFromCache(
+          explorationId);
+        if (successCallback) {
+          successCallback({});
```
You don't need to pass anything to ``successCallback``, ie just ``successCallback();`` is fine.

joshuacano replied:
Strangely enough the frontend test fails when I dont pass it something.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -0,0 +1,161 @@
+       * if one is provided to the returned promise object. Errors are passed
+       * to the error callback, if one is provided. Finally, if the update is
+       * successful, the returned exploration will be cached within the
+       * ReadOnlyExplorationBackendApiService to ensure the cache is
```
Is this true? Isn't the exploration removed from the read only backend api service if it's updated?

joshuacano replied:
Totally, this was an oversight on my part. Thanks for catching.

BenHenning commented on core/templates/dev/head/domain/exploration/ReadOnlyExplorationBackendApiService.js:
```diff
@@ -0,0 +1,163 @@
+       * with the exploration passed in as a parameter. If something goes
+       * wrong while trying to fetch the exploration, the rejection callback
+       * is called instead, if present. The rejection callback function is
+       * passed an error message using this format
```
How do we guarantee this format? From the earlier code, it seems we just pass whatever the backend tells us is the error. Is that not true?

joshuacano replied:
This is true, we do pass whatever the backend tells us is the error. In some cases it is the condition outlined in this comment, in others it is not. I will correct this comment.

BenHenning commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -85,28 +87,26 @@ oppia.factory('explorationData', [
+            explorationData.data = response;
 
-            return response.data;
+            return response;
```
Is this return value needed anymore?

joshuacano replied:
I certainly think so. It returns data for the getData method

BenHenning commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -85,28 +87,26 @@ oppia.factory('explorationData', [
-              apply_draft: true
-            }
-          }).then(function(response) {
+          return EditableExplorationBackendApiService.fetchExploration(
```
Does this return statement return the same thing it did before, or does some logic need to change now?

joshuacano replied:
It does do the same as before. logic should stay the same.

BenHenning commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -85,28 +87,26 @@ oppia.factory('explorationData', [
       // exploration.
       getLastSavedData: function() {
-        return $http.get(explorationDataUrl).then(function(response) {
+        return ReadOnlyExplorationBackendApiService.loadLatestExploration(
```
Ditto.

joshuacano replied:
I believe this also needs to return a promise with data as well. Surely the caller uses this data as well, no?

BenHenning commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -85,28 +87,26 @@ oppia.factory('explorationData', [
+          $log.info(response);
 
-          return response.data;
+          return response.exploration;
```
Ditto.

joshuacano replied:
Not much as changed here, are you thinking that I should change the promise behavior of this method? Or something else.

joshuacano replied:
I believe this should also return exploration to the promise as well.

BenHenning commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -85,28 +87,26 @@ oppia.factory('explorationData', [
-              apply_draft: true
-            }
-          }).then(function(response) {
+          return EditableExplorationBackendApiService.fetchExploration(
```
Why ``fetchExploration`` instead of load?

joshuacano replied:
In EditableExplorationBackendApiService there is only fetch as there is no cache ability. A bit different from the load capability in ReadOnlyExplorationBackendApiService which uses a cache. Does this seem ok to you?

BenHenning replied:
Yup, this makes sense now that I've taken a fresher look at this.

BenHenning commented on core/templates/dev/head/pages/exploration_editor/ExplorationEditor.js:
```diff
@@ -23,6 +23,15 @@ oppia.constant('PANEL_SPECS', GLOBALS.PANEL_SPECS);
+  '/explorehandler/init/<exploration_id>');
+oppia.constant(
+  'EXPLORATION_VERSION_DATA_URL_TEMPLATE',
+  '/explorehandler/init/<exploration_id>?v=<v>');
```
Prefer ``version`` over ``v`` to be more specific.

joshuacano replied:
Yep, makes total sense. Thanks for catching that.

joshuacano replied:
Great, I agree, this has been done.

BenHenning commented on core/templates/dev/head/pages/exploration_editor/history_tab/HistoryServicesSpec.js:
```diff
@@ -222,9 +224,9 @@ describe('Compare versions service', function() {
+        data.exploration.states[stateName] = newStateData;
       }
-      return statesData;
+      return data;
```
Rather than having an awkward ``data`` variable, you could keep the ``statesData`` variable and return a new object inline:

```javascript
return {
  exploration: {
    states: statesData
  }
};
```

joshuacano replied:
Yes! much better solution and much more understandable at a glance. This has been done.

joshuacano replied:
COMPLETELY, this is a great solution, much more readable.

BenHenning commented on core/templates/dev/head/pages/exploration_player/PlayerServices.js:
```diff
@@ -188,36 +190,33 @@ oppia.factory('oppiaPlayerService', [
-            }) + (version ? '?v=' + version : '');
-          $http.get(explorationDataUrl).then(function(response) {
-            var data = response.data;
+          loadedExploration = null;
```
Prefer keeping scope on this variable: ``var loadedExploration = null;``

joshuacano replied:
right. Fixed.

joshuacano replied:
Got it. done.

joshuacano commented:
Thanks Ben! Will address these tonight.

joshuacano commented:
@BenHenning`, apparently by tonight I meant 13 days from tonight. I believe I have addressed all review comments and have added 1 additional method to editableExplorationBackendApiService, titled fetchApplyDraftExploration. Review at your earliest convenience, thanks so much!

joshuacano commented:
@BenHenning Found why one of the tests was failing. Am fixing now and will have it committed shortly.

anmolshkl commented:
@BenHenning ping!

seanlip commented:
Hi @joshuacano -- sorry about this, but could you address the new merge conflict? Thanks!

joshuacano commented:
Addressed merge conflicts, Could not run e2e tests locally. Will check Travis in the morning to see if it failed.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -113,8 +93,36 @@ oppia.factory('EditableExplorationBackendApiService', [
+    var _getExplorationUrl = function(explorationId, applyDraft, role) {
+      if (role) {
+        return UrlInterpolationService.interpolateUrl(
+        EDITABLE_EXPLORATION_DATA_ROLE_URL_TEMPLATE, {
```
Is this spacing right? It seems like this would fail linting. I'd expect this and the next 3 lines to be indented rightward by 2, ditto for other similar ``return`` blocks.

joshuacano replied:
I would expect that as well, but the new linter gave me an error hence this strange format. Let me see if I can correct.

joshuacano replied:
The new linter passes the properly formatted lines. So long story short, this has been fixed, thanks!

BenHenning commented on core/templates/dev/head/pages/exploration_editor/ExplorationEditor.js:
```diff
@@ -28,10 +28,16 @@ oppia.constant(
   '/createhandler/data/<exploration_id>');
+oppia.constant(
+  'EDITABLE_EXPLORATION_DATA_DRAFT_URL_TEMPLATE',
+  '/createhandler/data/<exploration_id>?apply_draft=<applyDraft>');
```
``applyDraft`` should be ``apply_draft`` for consistency.

joshuacano replied:
This has been done.

BenHenning commented on core/templates/dev/head/pages/exploration_player/PlayerConstants.js:
```diff
@@ -27,7 +27,13 @@ oppia.constant(
+  '/createhandler/data/<exploration_id>');
+oppia.constant(
+  'EDITABLE_EXPLORATION_DATA_DRAFT_URL_TEMPLATE',
+  '/createhandler/data/<exploration_id>?apply_draft=<applyDraft>');
```
Ditto for this ``applyDraft``

joshuacano replied:
Totally, This has been done.

BenHenning commented on core/templates/dev/head/pages/exploration_player/PlayerServices.js:
```diff
@@ -212,11 +212,11 @@ oppia.factory('oppiaPlayerService', [
             exploration = ExplorationObjectFactory.createFromBackendDict(
               data.exploration);
-            version = exploration.version;
+            version = data.version;
```
Why change this from ``exploration.version``?

joshuacano replied:
This is because reader.py returns version outside of the exploration object. I did not notice that on my first pass.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -113,8 +93,36 @@ oppia.factory('EditableExplorationBackendApiService', [
+        });
+      },
+
+      fetchApplyDraftExploration: function(explorationId, applyDraft) {
```
Are there new tests for this new function?

joshuacano replied:
There is not, writing one now.

joshuacano replied:
I wrote a small, small test for this function. Ideally is the plan to add autosavedraft functionality in here as well? Or will that be a separate service. (I.e. EditableExplorationDraftBackendService)

BenHenning replied:
That's a good question and I haven't thought about it yet. Given the nature of these backend api services, we're sort of moving toward a 1:1 relationship between backend URLs and frontend api services, so I lean slightly toward a separate ``EditableExplorationDraftBackendApiService`` in the future.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiService.js:
```diff
@@ -145,12 +155,11 @@ oppia.factory('EditableExplorationBackendApiService', [
        */
-      deleteExploration: function(
-        explorationId, role) {
+      deleteExploration: function(explorationId, role) {
```
Do you ever test ``role`` in the frontend tests for this function?

joshuacano replied:
I will write a test that checks this.

joshuacano replied:
Actually, upon reviewing the code, Role is now handled in the backend and not passed as a variable anymore! See this [commit](https://github.com/oppia/oppia/commit/09279d02e6f6f197a3b9f1dc9be25e7d819b7df2).
 So will take out passing this variable from frontend, and remove associated URL templates.

BenHenning commented:
Thanks @joshuacano! I think this generally LGTM, though I'm curious about some missing tests (since you added some more functionality since I last took a look at this). One of the e2e tests failed and I restarted it (though it looks like it might be a real failure). It also seems you're failing the lint test, so please also fix any linting errors that are still outstanding.

seanlip commented:
@joshuacano, would you mind resolving the merge conflicts? Thanks!

joshuacano commented:
Writing a few new tests before I try and merge. The PR that will never die!
I should have some new commits soon.

On Tue, May 9, 2017 at 1:30 AM, Sean Lip <notifications@github.com> wrote:

> @joshuacano <https://github.com/joshuacano>, would you mind resolving the
> merge conflicts? Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/2999#issuecomment-300097349>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADKB7nexVnV9rJbEbeFInQy-PTSyDJgAks5r4CQPgaJpZM4LyJ7x>
> .
>

joshuacano commented:
I have addressed the review comments. If you'll take a look again @BenHenning, fingers crossed I think this might be the last change.

BenHenning commented on core/templates/dev/head/domain/exploration/EditableExplorationBackendApiServiceSpec.js:
```diff
@@ -90,6 +90,25 @@ describe('Editable exploration backend API service', function() {
     }
   );
 
+  it('should fetch and apply the draft of am exploration',
```
Nit: an instead of am.

joshuacano replied:
Fixed.

BenHenning commented on core/templates/dev/head/pages/exploration_editor/EditorServices.js:
```diff
@@ -20,9 +20,9 @@
   '$http', '$log', 'alertsService',
   'EditableExplorationBackendApiService',
-  'ReadOnlyExplorationBackendApiService', '$q',
+  'ReadOnlyExplorationBackendApiService','$q', 'UrlInterpolationService',
```
Why import ``UrlInterpolationService``? It isn't being used in this service from what I can tell.

joshuacano replied:
Must have been a hold over merge I missed. Thanks for spotting.

joshuacano commented:
ALL DONE! Thanks @BenHenning!

seanlip commented:
@joshuacano -- the e2e tests seem to be failing. Could you PTAL?

seanlip commented:
(From the logs, it looks like an issue with a missing dependency.)

joshuacano commented:
It was a missing dependency in the collection_editor, (Hadn't seen this before, it looks nice!)

This has been fixed.

seanlip commented:
Looks great, thanks @joshuacano! Merging.
